src/plugins/discover/public/application/main/services/discover_data_state_container.ts

```diff
@@ -0,0 +1,253 @@
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+import { BehaviorSubject, filter, map, Observable, share, Subject, tap } from 'rxjs';
+import { AutoRefreshDoneFn } from '@kbn/data-plugin/public';
+import { RequestAdapter } from '@kbn/inspector-plugin/common';
+import { SavedSearch } from '@kbn/saved-search-plugin/public';
+import { AggregateQuery, Query } from '@kbn/es-query';
+import type { SearchResponse } from '@elastic/elasticsearch/lib/api/types';
+import { ReduxLikeStateContainer } from '@kbn/kibana-utils-plugin/common';
+import { getRawRecordType } from '../utils/get_raw_record_type';
+import { AppState } from './discover_app_state_container';
+import { DiscoverServices } from '../../../build_services';
+import { DiscoverSearchSessionManager } from './discover_search_session';
+import { SEARCH_FIELDS_FROM_SOURCE, SEARCH_ON_PAGE_LOAD_SETTING } from '../../../../common';
+import { FetchStatus } from '../../types';
+import { validateTimeRange } from '../utils/validate_time_range';
+import { fetchAll } from '../utils/fetch_all';
+import { sendResetMsg } from '../hooks/use_saved_search_messages';
+import { getFetch$ } from '../utils/get_fetch_observable';
+import { DataTableRecord } from '../../../types';
+
+export interface SavedSearchData {
+  main$: DataMain$;
+  documents$: DataDocuments$;
+  totalHits$: DataTotalHits$;
+  availableFields$: AvailableFields$;
+}
+
+export type DataMain$ = BehaviorSubject<DataMainMsg>;
+export type DataDocuments$ = BehaviorSubject<DataDocumentsMsg>;
+export type DataTotalHits$ = BehaviorSubject<DataTotalHitsMsg>;
+export type AvailableFields$ = BehaviorSubject<DataAvailableFieldsMsg>;
+export type DataFetch$ = Observable<{
+  reset: boolean;
+  searchSessionId: string;
+}>;
+
+export type DataRefetch$ = Subject<DataRefetchMsg>;
+
+export enum RecordRawType {
+  /**
+   * Documents returned Elasticsearch, nested structure
+   */
+  DOCUMENT = 'document',
+  /**
+   * Data returned e.g. SQL queries, flat structure
+   * */
+  PLAIN = 'plain',
+}
+
+export type DataRefetchMsg = 'reset' | undefined;
+
+export interface DataMsg {
+  fetchStatus: FetchStatus;
+  error?: Error;
+  recordRawType?: RecordRawType;
+  query?: AggregateQuery | Query | undefined;
+}
+
+export interface DataMainMsg extends DataMsg {
+  foundDocuments?: boolean;
+}
+
+export interface DataDocumentsMsg extends DataMsg {
+  result?: DataTableRecord[];
+}
+
+export interface DataTotalHitsMsg extends DataMsg {
+  result?: number;
+}
+
+export interface DataChartsMessage extends DataMsg {
+  response?: SearchResponse;
+}
+
+export interface DataAvailableFieldsMsg extends DataMsg {
+  fields?: string[];
+}
+
+export interface DataStateContainer {
+  /**
+   * Implicitly starting fetching data from ES
+   */
+  fetch: () => void;
+  /**
+   * Observable emitting when a next fetch is triggered
+   */
+  fetch$: DataFetch$;
```

Comment:
I agree, fetch$ was just recently exposed for the histogram, so not exposing refetch$ (but just a `fetch` function which is more implicit) might make this clearer 

update: can't do this, because we need refetch$ here:
https://github.com/elastic/kibana/blob/2b3a5d2eec213631d2bef29c2558ee69d7547685/src/plugins/discover/public/application/main/components/field_stats_table/field_stats_table.tsx#L147)

So I think improving the description would be the way to go ... or naming 
`refetch$` -> `startFetch$` , `triggerFetch$`
open for suggestions 🙇